src/cas.rs

```diff
@@ -0,0 +1,111 @@
+//! Simple content-addressable file-based store.
+//!
+//! Give it an empty directory to save the files in,
+//! then call the `file_from_*` methods to write contents
+//! to the CAS. The content is hashed and a new file is only
+//! written if the content hasn’t been added before.
+//!
+//! Internally uses md5, don’t use for security-critical stuff.
+use std::io::Write;
+use std::path::PathBuf;
+
+/// A content-addressable store.
+#[derive(Clone)]
+pub struct ContentAddressable {
+    store_dir: PathBuf,
+}
+
+impl ContentAddressable {
+    /// Create a new content-addressable store.
+    ///
+    /// The `store_dir` *must* be an empty directory
+    /// (created if it does not yet exist), or a directory
+    /// pointing to another `ContentAddressable` store.
+    /// If it is a directory with some other data,
+    /// the correctness cannot be guaranteed.
+    pub fn new(store_dir: PathBuf) -> std::io::Result<ContentAddressable> {
+        std::fs::create_dir_all(&store_dir)?;
+        Ok(ContentAddressable { store_dir })
+    }
+
+    /// Adds the contents to a file in the content-addressable store
+    /// and returns the `PathBuf` pointing to the file.
+    ///
+    /// If the same contents were already written, the existing file
+    /// is returned.
+    ///
+    /// This operation hashes the file contents, its cost is at least
+    /// the cost of hashing the `content` string.
+    pub fn file_from_string(&self, content: &str) -> std::io::Result<PathBuf> {
+        // md5 should be okay, since this is not security-critical
+        let hash = format!("{:x}", md5::compute(content.as_bytes()));
+
+        let file_name = self.store_dir.join(hash);
+
+        match std::fs::OpenOptions::new()
+            .write(true)
+            .truncate(true)
+            .create_new(true)
+            .open(&file_name)
+        {
+            Err(e) => match e.kind() {
+                // since we use create_new, it will tell us if the file
+                // exists; in that case it was already written (same hash),
+                // so we don’t have to write it anew.
+                std::io::ErrorKind::AlreadyExists => Ok(()),
+                _ => Err(e),
+            },
+            Ok(mut file_handle) => {
+                file_handle.write_all(content.as_bytes())?;
+                file_handle.sync_all()
```

Comment:
If this gets interrupted, it is possible for the file to exist in the store with partial invalid content, and then never be rewritten again.

What if we did something like write to a temporary file and renamed the file in to place?

By having the temporary file in the same directory we are able to guarantee it is on the same filesystem, and thus guarantee that the `mv` is atomic.

Reply:
Another problem this write-and-move would solve is ensuring the file is only visible once it is fully written.

Reply:
https://github.com/grahamc/r13y.com/blob/master/src/contentaddressedstorage.rs#L28